ao.php

```diff
@@ -491,6 +491,44 @@ function ao_civicrm_alterAPIPermissions($entity, $action, &$params, &$permission
   }
 }
 
+/**
+ * Implements hook_civicrm_pre().
+ *
+ * AOS-64 Ensure that all postcodes are formatted in the format of xxx yyy
+ */
+function ao_civicrm_pre($op, $objectName, $id, &$params) {
+  $update = TRUE;
+  if ($objectName === 'Address' && ($op === 'create' || $op === 'edit')) {
+    if ($id) {
+      if (!empty($params['postal_code'])) {
+        $postalCode = str_replace(' ', '', $params['postal_code']);
+        $postalCode = substr($postalCode, 0, 3) . ' ' . substr($postalCode, 3);
+        if ($postalCode == $address['postal_code']) {
+          $update = FALSE;
+        }
+      }
+      else {
+        $address = civicrm_api3('Address', 'getsingle', ['id' => $id]);
```

Comment:
I'm not quite confident in this logic, but I'm not clear on all of the calling contexts for the function and the way that $id is set. Deciding to retain the old postal code when postal code is not entered but other elements of address are entered can lead to mismatch between postal code and rest of address.

Reply:
hook_civicrm_pre is called before database writes Joe, so for Address that is from https://github.com/civicrm/civicrm-core/blob/master/CRM/Core/BAO/Address.php#L132 . The id is only supplied if the calling function to essentially Address.create api or calling the create or add functions directly contains an ID and in that case the op is set to 'edit' rather than 'create' . So what i was trying to do here was to say ok if we have an id passed in lets look at the postcode in the database, clean it up and reformat it our way. If our reformat matches what is already in the database then we don't need to worry about updating the postal_code parameter